Number_types/include/CGAL/boost_mp_type.h

```diff
@@ -0,0 +1,1012 @@
+  typedef Residue Residue_type;
+
+  struct Modular_image{
+    Residue_type operator()(const NT& a){
+      NT tmp(CGAL::mod(a,NT(Residue::get_current_prime())));
+      return CGAL::Residue(tmp.template convert_to<int>());
+    }
+  };
+  struct Modular_image_representative{
+    NT operator()(const Residue_type& x){
+      return NT(x.get_value());
+    }
+  };
+};
+
+template <class Backend, boost::multiprecision::expression_template_option Eto>
+struct Modular_traits<boost::multiprecision::number<Backend, Eto> >
+: MT_boost_mp <boost::multiprecision::number<Backend, Eto> > {};
+template <class T1,class T2,class T3,class T4,class T5>
+struct Modular_traits<boost::multiprecision::detail::expression<T1,T2,T3,T4,T5> >
+: Modular_traits<typename boost::multiprecision::detail::expression<T1,T2,T3,T4,T5>::result_type > {};
+
+// Split_double
+
+template <class NT, class = boost::mpl::int_<boost::multiprecision::number_category<NT>::value> >
+struct SD_boost_mp {
+  void operator()(double d, NT &num, NT &den) const
+  {
+    num = d;
+    den = 1;
+  }
+};
+
+template <class NT>
+struct SD_boost_mp <NT, boost::mpl::int_<boost::multiprecision::number_kind_integer> >
+{
+  void operator()(double d, NT &num, NT &den) const
+  {
+    std::pair<double, double> p = split_numerator_denominator(d);
+    num = NT(p.first);
+    den = NT(p.second);
+  }
+};
+
+template <class Backend, boost::multiprecision::expression_template_option Eto>
+struct Split_double<boost::multiprecision::number<Backend, Eto> >
+: SD_boost_mp <boost::multiprecision::number<Backend, Eto> > {};
+template <class T1,class T2,class T3,class T4,class T5>
+struct Split_double<boost::multiprecision::detail::expression<T1,T2,T3,T4,T5> >
+: Split_double<typename boost::multiprecision::detail::expression<T1,T2,T3,T4,T5>::result_type > {};
+
+
+// Fraction_traits
+
+template <class T, class = boost::mpl::int_<boost::multiprecision::number_category<T>::value> >
+struct FT_boost_mp {
+  typedef T Type;
+  typedef Tag_false Is_fraction;
+  typedef Null_tag Numerator_type;
+  typedef Null_tag Denominator_type;
+  typedef Null_functor Common_factor;
+  typedef Null_functor Decompose;
+  typedef Null_functor Compose;
+};
+
+template <class NT>
+struct FT_boost_mp <NT, boost::mpl::int_<boost::multiprecision::number_kind_rational> > {
+    typedef NT Type;
+
+    typedef ::CGAL::Tag_true Is_fraction;
+    typedef typename boost::multiprecision::component_type<NT>::type Numerator_type;
+    typedef Numerator_type Denominator_type;
+
+    typedef typename Algebraic_structure_traits< Numerator_type >::Gcd Common_factor;
+
+    class Decompose {
+    public:
+        typedef Type first_argument_type;
+        typedef Numerator_type& second_argument_type;
+        typedef Denominator_type& third_argument_type;
+        void operator () (
+                const Type& rat,
+                Numerator_type& num,
+                Denominator_type& den) {
+            num = numerator(rat);
```

Comment:
@sloriot don't we have a problem here, namely that we do not get a reference of the numerator and denominator?

Reply:
No, a copy is made and put into the reference passed as input.